lib/honeybadger/insights/base.ex

```diff
@@ -0,0 +1,164 @@
+defmodule Honeybadger.Insights.Base do
+  @moduledoc """
+  Base module providing common telemetry attachment functionality.
+  """
+
+  defmacro __using__(_opts) do
+    quote do
+      require Logger
+
+      # Define empty defaults for the attributes
+      Module.register_attribute(__MODULE__, :required_dependencies, [])
+      @required_dependencies []
+      Module.register_attribute(__MODULE__, :telemetry_events, [])
+      @time_keys ~w(duration total_time decode_time query_time queue_time idle_time)a
+
+      @before_compile Honeybadger.Insights.Base
+
+      def dependencies_available? do
+        Honeybadger.Insights.Base.dependencies_available?(@required_dependencies)
+      end
+
+      def get_insights_config(key, default) do
+        Honeybadger.Insights.Base.get_insights_config(key, default, __MODULE__)
+      end
+    end
+  end
+
+  @doc """
+  Checks if all required dependencies are available.
+  """
+  def dependencies_available?(deps) do
+    if Enum.empty?(deps), do: true, else: Enum.all?(deps, &Code.ensure_loaded?/1)
+  end
+
+  @doc """
+  Retrieves a configuration value from the insights configuration.
+  """
+  def get_insights_config(key, default, mod) do
+    config_namespace =
+      mod
+      |> Atom.to_string()
+      |> String.split(".")
+      |> List.last()
+      |> Macro.underscore()
+      |> String.to_atom()
+
+    insights_config = Application.get_env(:honeybadger, :insights_config, %{})
+    module_config = Map.get(insights_config, config_namespace, %{})
+    Map.get(module_config, key, default)
+  end
+
+  defmacro __before_compile__(_env) do
+    quote do
+      def event_filter(map, meta, name) do
+        if Application.get_env(:honeybadger, :event_filter) do
+          Application.get_env(:honeybadger, :event_filter).filter(map, meta, name)
+        else
+          map
+        end
+      end
+
+      def get_telemetry_events() do
+        get_insights_config(:telemetry_events, @telemetry_events)
+      end
+
+      @doc """
+      Attaches telemetry handlers if all required dependencies are available.
+      """
+      def attach do
+        if dependencies_available?() do
+          get_telemetry_events()
+          |> Enum.each(&attach_event/1)
+
+          :ok
+        else
+          Logger.debug(
+            "[Honeybadger] Missing Insights dependencies for #{inspect(@required_dependencies)}"
+          )
+
+          {:error, :missing_dependencies}
+        end
+      end
+
+      @doc """
+      Attaches a telemetry handler for a specific event.
+      The event can be either a string (dot-separated) or a list of atoms.
+      """
+      def attach_event(event) when is_binary(event) do
+        event_atoms = event |> String.split(".") |> Enum.map(&String.to_atom/1)
```

Comment:
Use of `String.to_atom/1` isn't considered safe and this will get flagged by security scanners. Use `to_existing_atom/1` instead

```suggestion
        event_atoms = event |> String.split(".") |> Enum.map(&String.to_existing_atom/1)
```

Reply:
I made this change but I had to add some atoms to the start of the test_helper, as they were not available. 

@sorentwo This makes me wonder if using `to_existing_atom` could raise errors for some users, perhaps if they have an early version of a library that does not include the atoms used in the telemetry subscription? What if we just ignore the credo check with a prepending comment, as the strings should only be coming from the module var or the application config.